### PR TITLE
ffmpeg-full: better support for cross-compilation to windows, add openh264

### PR DESCRIPTION
**Pull request by marius851000**
###### Motivation for this change
I needed to have a custom ffmpeg executable for windows.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

The option I used to build it can be found here :

https://gist.github.com/marius851000/00bc703cc8031bb9195f87f51556346b

In particular :

- I made windres use the appropriate compiler using another method that take the absolute path of the executable, remove the path (keep only the binary name), strip up to the last dash, then add the compiler name, and use it from the PATH (so, it'll use ``x86_64-w64-mingw32-gcc`` instead of just ``gcc``, which are not present in this context). This is necessary at least for gettext and glib.
- other change to get the minimal necessary stuff to build ffmpeg to compile on windows
- added openh264 to ffmpeg.

*also, the way DLL work is ugly*

### CODE REVIEW

File: pkgs/development/libraries/ffmpeg-full/default.nix [hunk 1]
```diff
@@ -259,7 +259,7 @@ stdenv.mkDerivation rec {
 
   configurePlatforms = [];
   configureFlags = [
-    "--target_os=${stdenv.hostPlatform.parsed.kernel.name}"
+    "--target_os=${if stdenv.hostPlatform.isWindows then "mingw64" else stdenv.hostPlatform.parsed.kernel.name}" #mingw32 and mingw64 doesn't have a difference here, it is internally rewritten as mingw32
```
Comment (Ericson2314) [inline pkgs/development/libraries/ffmpeg-full/default.nix:262]:
```suggestion
    "--target_os=${if stdenv.hostPlatform.isMinGW then "mingw64" else stdenv.hostPlatform.parsed.kernel.name}" #mingw32 and mingw64 doesn't have a difference here, it is internally rewritten as mingw32
```

File: pkgs/development/libraries/glib/default.nix [hunk 2]
```diff
@@ -93,11 +93,12 @@ stdenv.mkDerivation rec {
...
+  ] ++ (optionals (!stdenv.hostPlatform.isWindows) [
     bash gnum4 # install glib-gettextize and m4 macros for other apps to use
-  ] ++ optionals stdenv.isLinux [
+  ]) ++ (optionals stdenv.isLinux [
     libselinux
     util-linuxMinimal # for libmount
-  ] ++ optionals stdenv.isDarwin (with darwin.apple_sdk.frameworks; [
+  ]) ++ optionals stdenv.isDarwin (with darwin.apple_sdk.frameworks; [
     AppKit Carbon Cocoa CoreFoundation CoreServices Foundation
   ]) ++ optionals (stdenv.hostPlatform == stdenv.buildPlatform) [
```
Comment (Ericson2314) [inline pkgs/development/libraries/glib/default.nix:103]: The parens around the  `(optionals ... [ ... ])` are not needed.

File: pkgs/development/libraries/glib/default.nix [hunk 3]
```diff
@@ -143,7 +144,10 @@ stdenv.mkDerivation rec {
     patchShebangs glib/gen-unicode-tables.pl
     patchShebangs tests/gen-casefold-txt.py
     patchShebangs tests/gen-casemap-txt.py
-  '';
+  '' + (if stdenv.hostPlatform.isWindows then ''
+    substituteInPlace gio/win32/meson.build \
+      --replace "libintl, " ""
+  '' else "");
```
Comment (Ericson2314) [inline pkgs/development/libraries/glib/default.nix:150]:
```suggestion
  '' + lib.optionalString stdenv.hostPlatform.isWindows ''
    substituteInPlace gio/win32/meson.build \
      --replace "libintl, " ""
  '';
```

File: pkgs/development/libraries/openh264/default.nix [hunk 4]
```diff
@@ -13,10 +13,16 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = [ nasm ];
 
+  buildInputs = if stdenv.hostPlatform.isWindows then [ windows.pthreads ] else [];
```
Comment (Ericson2314) [inline pkgs/development/libraries/openh264/default.nix:16]:
```suggestion
  buildInputs = lib.optional stdenv.hostPlatform.isWindows windows.pthreads;
```

File: pkgs/development/tools/misc/binutils/windres-locate-gcc.patch [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
++      /* Nixpkgs specific : look first at the prefixed path
++        ( there should be no gcc in the binutils folder ) */
++      if (slash && dash) {
++        cpp_pipe = look_for_default(cmd, slash + 1, dash - slash, preprocargs, filename);
++      }
++
++      if (dash && ! cpp_pipe)
+ 	{
+ 	  /* First, try looking for a prefixed gcc in the windres
+ 	     directory, with the same prefix as windres */
```
Comment (Ericson2314) [inline pkgs/development/tools/misc/binutils/windres-locate-gcc.patch:19]: What happens without this?

Comment (marius851000) [reply pkgs/development/tools/misc/binutils/windres-locate-gcc.patch:19]: The original code for this part was

```C
	if (dash)
	{
	  /* First, try looking for a prefixed gcc in the windres
	     directory, with the same prefix as windres */

	  cpp_pipe = look_for_default (cmd, program_name, dash - program_name + 1,
				       preprocargs, filename);
	}

      if (slash && ! cpp_pipe)
	{
	  /* Next, try looking for a gcc in the same directory as
             that windres */

	  cpp_pipe = look_for_default (cmd, program_name, slash - program_name + 1,
				       preprocargs, filename);
	}

      if (! cpp_pipe)
	{
	  /* Sigh, try the default */

	  cpp_pipe = look_for_default (cmd, "", 0, preprocargs, filename);
	}
```


The cpp_pipe is pointer to an open file (I think it's the gcc binary).


slash contain a pointer to the last ``/`` in the filename in memory, and dash a pointer to the last ``-``.

Then, it look for gcc (without this patch) in 3 different place (ordered by search order):

(where COMMAND is the command entered in the terminal)

if there is a dash, <COMMAND up to the last dash>
if not found and slash, <COMMAND up to the last slash>
default: <nothing> (seemingly equivalent to look for unprefixed gcc from PATH)

I added a fourth possibility, that will be tried first :
if there is a slash and a dash, then the binary was probably called with it's full name.
we cut the between the last slash and the last dash. That does something like (for COMMAND)
``a/b/c/something-with-dash`` will try with ``something-with-``.

Comment (symphorien) [reply pkgs/development/tools/misc/binutils/windres-locate-gcc.patch:19]: The logic you describe seems reasonable, even outside nixpkgs, maybe send the patch upstream ?

Comment (marius851000) [reply pkgs/development/tools/misc/binutils/windres-locate-gcc.patch:19]: I will do this eventually. I'm just a bit intimitided by the absence of gitlab/whatever for PR (guess I should send this on the bug mailing list). And the fact that ``nix develop nixpkgs#binutils`` doesn't work.

**Pull request merged**